Config/config.txt
{
	"debug": "true",
	
	"bot": {
		"name": "snackbot",
		"alt_name": "Snack-Bot",
		"login": "itsabot",
		"version": "maybeitllcrashless",
		"password": "imnotgivingawaymypass"
	},
	
	"server": "irc.rizon.net",
	
	"channels": {
		"channel1": {
          		"name": "#pantsumen",
            		"html": "true",
            		"lastfm": "true",
            		"weather": "true",
            		"quote": "false",
            		"tell": "true"
		}
	},
	
	"settings": {
		"config": "E:\\ircbot\\"
	},
	
	"disable": {
		"bot1": {
			"name": "uguubot",
			"html": "false",
			"lastfm": "true",
			"weather": "true",
			"quote": "false",
			"tell": "true"
		},
		"bot2": {
			"name": "MikuBot",
			"html": "true",
			"lastfm": "false",
			"weather": "false",
			"quote": "false",
			"tell": "false"
		}
	}
}


IrcBot/config.txt
{
	"admin": "snack",
	"debug": "false",
	"interface": "true",
	"scrollback": "50",
	
	"bot": {
		"name": "snackbot",
		"alt_name": "Snack-Bot",
		"login": "itsabot",
		"version": "maybeitllcrashless",
		"password": "1.pieps"
	},
	
	"server": "irc.rizon.net",
	
	"channels": {
		"channel1": {
			"name": "#pantsumen",
			"functions": {
				"greet": {
					"enabled": "false",
					"visible": "public"
				},
	            "html": "true",
	            "lastfm": "true",
	            "weather": "true",
	            "quote": "true",
	            "tell": "true",
				"translate": "false",
				"romaji": "false",
				"search": {
					"enabled": "true",
					"default": "g"
				}
			},
			"bots": {
				"bot1": {
					"name": "uguubot",
					"greet": "false",
					"html": "false",
					"lastfm": "true",
					"weather": "true",
					"quote": "false",
					"tell": "true",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				},
				"bot2": {
					"name": "MikuBot",
					"greet": "false",
					"html": "true",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				},
				"bot3": {
					"name": "rinpantsu",
					"greet": "true",
					"html": "false",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "true",
					"search": "false"
				},
				"bot4": {
					"name": "Moelicious",
					"greet": "false",
					"html": "false",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				}
			}
		},
		"channel2": {
			"name": "#rice",
			"functions": {
				"greet": {
					"enabled": "true",
					"visible": "public"
				},
		        "html": "true",
		        "lastfm": "true",
		        "weather": "true",
		        "quote": "false",
		        "tell": "true",
				"translate": "false",
				"romaji": "false",
				"search": {
					"enabled": "true",
					"default": "g"
				}
			},
			"bots": {
				"bot1": {
					"name": "uguubot",
					"greet": "false",
					"html": "true",
					"lastfm": "true",
					"weather": "true",
					"quote": "false",
					"tell": "true",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				},
				"bot2": {
					"name": "Moelicious",
					"greet": "true",
					"html": "false",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				}
			}
		},
		"channel3": {
			"name": "#/fit/",
			"functions": {
				"greet": {
					"enabled": "false",
					"visible": "public"
				},
		        "html": "true",
		        "lastfm": "true",
		        "weather": "true",
		        "quote": "false",
		        "tell": "true",
				"translate": "false",
				"romaji": "false",
				"search": {
					"enabled": "false",
					"default": "fit"
				}
			},
			"bots": {
				"bot1": {
					"name": "NBA",
					"greet": "false",
					"html": "false",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				},
				"bot2": {
					"name": "godzilla",
					"greet": "true",
					"html": "false",
					"lastfm": "false",
					"weather": "false",
					"quote": "false",
					"tell": "false",
					"translate": "false",
					"romaji": "false",
					"search": "false"
				}
			}
		}
	},
	
	"settings": {
		"config": "E:\\ircbot\\",
		"identifiers": ".,!",
		"quote": {
			"min": "10",
			"max": "150"
		}
	},
	
	"speech": {
		"mute": {
			"mute":	"I'll be silent.",
			"unmute": "Yay! I can speak again."
		},
		"weather": {
			"success": "<city>: <condition> <degrees_c>°C/ <degrees_f>°F Wind: <wind_dir> at <wind_speed> <wind_type> gusting to <wind_speed_gust> <wind_type> Humidity: <humidity>",
			"error": "No data found for the given location, or too many request made to the API."
		},
		"lastfm": {
			"success": {
				"now_playing": "<username> is now playing: <song> by <artist> on <album>",
				"last_played": "<username> last played: <song> by <artist> on <album>",
				"no_artist": "Unknown Artist",
				"no_album": "Unknown Album",
				"no_song": "Unknown Song"
			},
			"error": "No data found for the given name."
		},
		"greet": "Ohayou <name>",
		"html": {
			"success": {
				"title": "[URI] <title>",
				"file": {
					"size_zero": "Unknown size",
					"size_not_zero": "<size> <format>",
					"encoding": "'<content-type>, <content-encoding>'",
					"no_encoding": "'<content-type>'",
					"file_full": "<encoding> <size>"
				}
			},
			"error": {
				"403": "403 Forbidden",
				"404": "404 Not Found",
				"?": "No title or file found."
			}
		},
		"quote": {
			"success": "<<name>> <quote>",
			"error": "No quote found.",
			"add": "Quote added."
		},
		"tell": {
			"add": "I'll tell that to <name> for you.",
			"tell": "<sender> sent you a message: <message>",
			"error": "<name> is online, do it yourself."
		},
		"translate": {
			"success": "Translated: <response>",
			"error": "Something went wrong translating."
		},
		"romaji": {
			"romaji": "Romaji: <response>",
			"katakana": "Katakana: <response>"
		},
		"search": {
			"success": "Thread: <subject> > <url>",
			"error": "Nothing found."
		}
	}
}


IrcBot/src/org/snack/irc/settings/Config.java
package org.snack.irc.settings;

import java.util.ArrayList;
import java.util.HashMap;

import net.sf.json.JSONObject;
import net.sf.json.JSONSerializer;

import org.snack.irc.model.Bot;
import org.snack.irc.model.Chan;

public class Config {
	private static Config c;
	private static JSONObject jo;

	/**
	 * Make sure it only gets read once, reads a config file and stores the data
	 */
	public static void initialize() throws Exception {
		if (c == null) {
			c = new Config();
		}
		jo = (JSONObject) JSONSerializer.toJSON(TxtReader.parseTxt("config.txt"));

		initSettings();
		initSpeech();
	}

	// All the settings
	public static HashMap<String, String> sett_str;
	public static HashMap<String, Boolean> sett_bool;
	public static HashMap<String, Integer> sett_int;
	public static HashMap<String, Chan> channels;

	/**
	 * Initialize the settings
	 */
	private static void initSettings() {
		sett_str = new HashMap<String, String>();
		sett_bool = new HashMap<String, Boolean>();
		sett_int = new HashMap<String, Integer>();
		channels = new HashMap<String, Chan>();

		sett_str.put("ADMIN", jo.getString("admin"));
		sett_bool.put("DEBUG", jo.getBoolean("debug"));
		sett_bool.put("INTERFACE", jo.getBoolean("interface"));
		sett_int.put("SCROLLBACK", jo.getInt("scrollback"));

		JSONObject joResultBot = jo.getJSONObject("bot");
		sett_str.put("BOT_NAME", joResultBot.getString("name"));
		sett_str.put("BOT_ALT_NAME", joResultBot.getString("alt_name"));
		sett_str.put("BOT_LOGIN", joResultBot.getString("login"));
		sett_str.put("BOT_VERSION", joResultBot.getString("version"));
		sett_str.put("BOT_PASS", joResultBot.getString("password"));

		sett_str.put("SERVER", jo.getString("server"));

		JSONObject joResultChannels = jo.getJSONObject("channels");
		for (int i = 1; i <= joResultChannels.size(); i++) {
			JSONObject obj = joResultChannels.getJSONObject("channel" + i);
			JSONObject func = obj.getJSONObject("functions");
			JSONObject greet = func.getJSONObject("greet");
			JSONObject search = func.getJSONObject("search");
			JSONObject joResultBots = obj.getJSONObject("bots");
			ArrayList<Bot> bots = new ArrayList<Bot>();
			for (int j = 1; j <= joResultBots.size(); j++) {
				JSONObject bot = joResultBots.getJSONObject("bot" + j);
				bots.add(new Bot(bot.getString("name"), bot.getBoolean("greet"), bot.getBoolean("html"), bot.getBoolean("lastfm"), bot.getBoolean("weather"), bot
						.getBoolean("quote"), bot.getBoolean("tell"), bot.getBoolean("translate"), bot.getBoolean("romaji"), bot.getBoolean("search")));
			}
			Chan chan = new Chan(obj.getString("name"), greet.getBoolean("enabled"), greet.getString("visible"), func.getBoolean("html"), func.getBoolean("lastfm"),
					func.getBoolean("weather"), func.getBoolean("quote"), func.getBoolean("tell"), func.getBoolean("translate"), func.getBoolean("romaji"),
					search.getBoolean("enabled"), search.getString("default"), bots);
			channels.put(chan.name, chan);
		}

		JSONObject joResultSettings = jo.getJSONObject("settings");
		sett_str.put("SAVE_LOC", joResultSettings.getString("config"));
		sett_str.put("IDENTIFIERS", joResultSettings.getString("identifiers"));
		JSONObject joQuoteSettings = joResultSettings.getJSONObject("quote");
		sett_int.put("QUOTE_MIN", joQuoteSettings.getInt("min"));
		sett_int.put("QUOTE_MAX", joQuoteSettings.getInt("max"));
	}

	// All the speech lines
	public static HashMap<String, String> speech;

	/**
	 * Initialize the speech lines
	 */
	private static void initSpeech() {
		speech = new HashMap<String, String>();

		JSONObject joSpeech = jo.getJSONObject("speech");

		JSONObject joMute = joSpeech.getJSONObject("mute");
		speech.put("MUTE", joMute.getString("mute"));
		speech.put("UNMUTE", joMute.getString("unmute"));

		speech.put("GREET", joSpeech.getString("greet"));

		JSONObject joWeather = joSpeech.getJSONObject("weather");
		speech.put("WE_SUC", joWeather.getString("success"));
		speech.put("WE_ERR", joWeather.getString("error"));

		JSONObject joLastfm = joSpeech.getJSONObject("lastfm");
		JSONObject joLastfmSuccess = joLastfm.getJSONObject("success");
		speech.put("LA_SUC_NP", joLastfmSuccess.getString("now_playing"));
		speech.put("LA_SUC_LP", joLastfmSuccess.getString("last_played"));
		speech.put("LA_SUC_ART", joLastfmSuccess.getString("no_artist"));
		speech.put("LA_SUC_ALB", joLastfmSuccess.getString("no_album"));
		speech.put("LA_SUC SON", joLastfmSuccess.getString("no_song"));
		speech.put("LA_ERR", joLastfm.getString("error"));

		JSONObject joHtml = joSpeech.getJSONObject("html");
		JSONObject joHtmlSuccess = joHtml.getJSONObject("success");
		JSONObject joHtmlSuccessFile = joHtmlSuccess.getJSONObject("file");
		JSONObject joHtmlError = joHtml.getJSONObject("error");
		speech.put("HT_TIT", joHtmlSuccess.getString("title"));
		speech.put("HT_SUC_SIZE0", joHtmlSuccessFile.getString("size_zero"));
		speech.put("HT_SUC_!SIZE0", joHtmlSuccessFile.getString("size_not_zero"));
		speech.put("HT_SUC_ENC", joHtmlSuccessFile.getString("encoding"));
		speech.put("HT_SUC_!ENC", joHtmlSuccessFile.getString("no_encoding"));
		speech.put("HT_SUC", joHtmlSuccessFile.getString("file_full"));
		speech.put("HT_ERR_403", joHtmlError.getString("403"));
		speech.put("HT_ERR_404", joHtmlError.getString("404"));
		speech.put("HT_ERR_?", joHtmlError.getString("?"));

		JSONObject joQuote = joSpeech.getJSONObject("quote");
		speech.put("QU_SUC", joQuote.getString("success"));
		speech.put("QU_ERR", joQuote.getString("error"));
		speech.put("QU_ADD", joQuote.getString("add"));

		JSONObject joTell = joSpeech.getJSONObject("tell");
		speech.put("TE_ADD", joTell.getString("add"));
		speech.put("TE_TEL", joTell.getString("tell"));
		speech.put("TE_ERR", joTell.getString("error"));

		JSONObject joTranslate = joSpeech.getJSONObject("translate");
		speech.put("TR_SUC", joTranslate.getString("success"));
		speech.put("TR_ERR", joTranslate.getString("error"));

		JSONObject joRomaji = joSpeech.getJSONObject("romaji");
		speech.put("RK_ROM", joRomaji.getString("romaji"));
		speech.put("RK_KAT", joRomaji.getString("katakana"));

		JSONObject joSearch = joSpeech.getJSONObject("search");
		speech.put("SE_SUC", joSearch.getString("success"));
		speech.put("SE_ERR", joSearch.getString("error"));
	}
}
